### PR TITLE
Improve cluster failure testing in 2.2

### PR DESCRIPTION
**Pull request by rickardoberg**
"Improve cluster failure testing" for 2.2: https://github.com/neo4j/neo4j/pull/4537

### CODE REVIEW

File: community/kernel/src/test/java/org/neo4j/kernel/lifecycle/LifeRule.java [hunk 1]
```diff
@@ -52,21 +53,50 @@ public void evaluate() throws Throwable
...
                     {
-                        shutdown();
+                        life.shutdown();
                     }
                     catch ( Throwable suppressed )
                     {
                         failure.addSuppressed( suppressed );
                     }
                     throw failure;
+                } finally
```
Comment (davidegrohmann) [inline community/kernel/src/test/java/org/neo4j/kernel/lifecycle/LifeRule.java]: code style

File: enterprise/cluster/src/main/java/org/neo4j/cluster/StateMachines.java [hunk 2]
```diff
@@ -190,6 +190,8 @@ public void run()
                                 }
                                 else
                                 {
+//                                    message.copyHeadersTo( outgoingMessage, Message.INSTANCE_ID );
+
```
Comment (davidegrohmann) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/StateMachines.java]: to be removed?

File: enterprise/cluster/src/main/java/org/neo4j/cluster/com/NetworkReceiver.java [hunk 3]
```diff
@@ -212,22 +219,25 @@ public void addMessageProcessor( MessageProcessor processor )
 
     public void receive( Message message )
     {
-        for ( MessageProcessor processor : processors )
+        if (!paused)
```
Comment (davidegrohmann) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/com/NetworkReceiver.java]: code style

Comment (davidegrohmann) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/com/NetworkReceiver.java]: perhaps this is more readable written like:

```
if ( paused )
{
    return;
}

...
```

File: enterprise/cluster/src/main/java/org/neo4j/cluster/com/NetworkSender.java [hunk 4]
```diff
@@ -215,18 +217,26 @@ public void process( final List<Message<? extends MessageType>> messages )
     @Override
     public boolean process( Message<? extends MessageType> message )
     {
-        if ( message.hasHeader( Message.TO ) )
-        {
-            send( message );
-        }
-        else
+        if (!paused)
```
Comment (davidegrohmann) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/com/NetworkSender.java]: code style. Invert if and return early?

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/AcceptorMessage.java [hunk 5]
```diff
@@ -145,7 +148,20 @@ public int hashCode()
...
+            Object toStringValue = value;
+            if (toStringValue instanceof Payload )
+            {
+                try
+                {
+                    toStringValue = new AtomicBroadcastSerializer( new ObjectStreamFactory(), new ObjectStreamFactory() ).receive( (Payload) toStringValue);
+                }
+                catch ( Throwable e )
+                {
+                    e.printStackTrace();
```
Comment (davidegrohmann) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/AcceptorMessage.java]: what's the benefit of printing the stack trace?

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/InMemoryAcceptorInstanceStore.java [hunk 6]
```diff
@@ -51,6 +56,11 @@ public AcceptorInstance getAcceptorInstance( InstanceId instanceId )
         {
             instance = new AcceptorInstance();
             instances.put( instanceId, instance );
+            if (!currentInstances.offer( instanceId ))
```
Comment (davidegrohmann) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/InMemoryAcceptorInstanceStore.java]: code style

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/InMemoryAcceptorInstanceStore.java [hunk 7]
```diff
@@ -87,7 +92,9 @@ public void clear()
 
     public InMemoryAcceptorInstanceStore snapshot()
     {
-        return new InMemoryAcceptorInstanceStore( new HashMap<>(instances), lastDeliveredInstanceId );
+        return new InMemoryAcceptorInstanceStore( new HashMap<>(instances),
+                new ArrayBlockingQueue<>( currentInstances.size()+currentInstances.remainingCapacity(), false, currentInstances ),
```
Comment (davidegrohmann) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/InMemoryAcceptorInstanceStore.java]: code style

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/LearnerMessage.java [hunk 8]
```diff
@@ -50,6 +53,18 @@ public Object getValue()
         @Override
         public String toString()
         {
+            if (value instanceof Payload )
```
Comment (davidegrohmann) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/LearnerMessage.java]: code style

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/LearnerMessage.java [hunk 9]
```diff
@@ -50,6 +53,18 @@ public Object getValue()
...
+            {
+                try
+                {
+                    ObjectStreamFactory streamFactory = new ObjectStreamFactory();
+                    return new AtomicBroadcastSerializer( streamFactory, streamFactory ).receive( (Payload) value).toString();
+                }
+                catch ( Throwable e )
+                {
+                    return value.toString();
```
Comment (davidegrohmann) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/LearnerMessage.java]: not needed

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/LearnerMessage.java [hunk 10]
```diff
@@ -50,6 +53,18 @@ public Object getValue()
         @Override
         public String toString()
         {
+            if (value instanceof Payload )
+            {
+                try
+                {
+                    ObjectStreamFactory streamFactory = new ObjectStreamFactory();
+                    return new AtomicBroadcastSerializer( streamFactory, streamFactory ).receive( (Payload) value).toString();
```
Comment (davidegrohmann) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/LearnerMessage.java]: To avoid duplication of code, why don't move this logic into `Payload.toString`?

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/LearnerState.java [hunk 11]
```diff
@@ -271,12 +273,13 @@ private URI lastKnownAliveUriOrSenderUri( LearnerContext context, Message<Learne
                         throws URISyntaxException
                 {
                     org.neo4j.cluster.InstanceId lastKnownAliveInstance = context.getLastKnownAliveUpToDateInstance();
-                    if ( lastKnownAliveInstance == null )
+                    if ( lastKnownAliveInstance != null )
                     {
-                        lastKnownAliveInstance =
-                                context.getIdForUri( new URI( message.getHeader( Message.FROM ) ) );
+                        return context.getUriForId( lastKnownAliveInstance );
+                    } else
```
Comment (davidegrohmann) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/LearnerState.java]: code style

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/PaxosInstance.java [hunk 12]
```diff
@@ -191,10 +199,43 @@ public PaxosInstance snapshot(PaxosInstanceStore store)
...
+
+            return "[id:" + id +
+                   " state:" + state.name() +
+                   " b:" + ballot +
+                   " v1:" + toStringValue1 +
+                   " v2:" + toStringValue2 + "]";
+        }
+        catch ( Throwable e )
+        {
+            return "";
```
Comment (davidegrohmann) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/PaxosInstance.java:238]: can't we do better than this? perhaps skipping only the values causing the issues?

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/heartbeat/HeartbeatState.java [hunk 13]
```diff
@@ -229,8 +229,8 @@ public HeartbeatState handle( HeartbeatContext context,
                             context.getLogger( HeartbeatState.class )
                                     .debug( "Received suspicions as " + suspicions );
 
-                            URI from = new URI( message.getHeader( Message.FROM ) );
-                            InstanceId fromId = context.getIdForUri( from );
+                            InstanceId fromId = new InstanceId(Integer.parseInt(message.getHeader( Message.INSTANCE_ID )));
```
Comment (davidegrohmann) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/heartbeat/HeartbeatState.java]: code style

File: enterprise/cluster/src/test/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/ProposerStateTest.java [hunk 14]
```diff
@@ -90,6 +92,36 @@ public void ifProposingWithClosedInstanceThenRetryWithNextInstance() throws Thro
     }
 
     @Test
+    public void something() throws Throwable
```
Comment (davidegrohmann) [inline enterprise/cluster/src/test/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/ProposerStateTest.java]: something?

File: enterprise/cluster/src/test/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/ProposerStateTest.java [hunk 15]
```diff
@@ -90,6 +92,36 @@ public void ifProposingWithClosedInstanceThenRetryWithNextInstance() throws Thro
...
+                        create( "http://something3" ) ) ) );
+
+        Message message = Message.to( ProposerMessage.promise, create( "http://something1" ), new ProposerMessage.PromiseState( 2001, acceptorValue ) );
+        message.setHeader( org.neo4j.cluster.protocol.atomicbroadcast.multipaxos.InstanceId.INSTANCE, instanceId.toString() );
+
+        MessageHolder mockHolder = mock( MessageHolder.class );
+        ProposerState.proposer.handle(context, message, mockHolder);
+
+
+
```
Comment (davidegrohmann) [inline enterprise/cluster/src/test/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/ProposerStateTest.java:121]: no assertions?

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/HighAvailabilityModeSwitcher.java [hunk 16]
```diff
@@ -349,6 +349,7 @@ public void run()
                     ClusterClient clusterClient = dependencyResolver.resolveDependency( ClusterClient.class );
                     try
                     {
+                        // TODO I doubt this actually works
```
Comment (davidegrohmann) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/HighAvailabilityModeSwitcher.java]: can you explain why?

File: enterprise/ha/src/test/java/org/neo4j/ha/UniquenessConstraintValidationHAIT.java [hunk 17]
```diff
@@ -86,33 +96,43 @@ public void shouldAllowCreationOfNonConflictingDataOnSeparateHosts() throws Exce
         assertTrue( "creating non-conflicting data should pass", created.get() );
     }
 
+    @RepeatRule.Repeat(times=1000)
```
Comment (davidegrohmann) [inline enterprise/ha/src/test/java/org/neo4j/ha/UniquenessConstraintValidationHAIT.java]: to be removed?

File: enterprise/ha/src/test/java/org/neo4j/ha/UniquenessConstraintValidationHAIT.java [hunk 18]
```diff
@@ -86,33 +96,43 @@ public void shouldAllowCreationOfNonConflictingDataOnSeparateHosts() throws Exce
...
     @Test
     public void shouldPreventConcurrentCreationOfConflictingDataOnSeparateHosts() throws Exception
     {
+        System.out.println(repeat.getCount());
```
Comment (davidegrohmann) [inline enterprise/ha/src/test/java/org/neo4j/ha/UniquenessConstraintValidationHAIT.java]: to be removed?

File: enterprise/ha/src/test/java/org/neo4j/kernel/ha/ClusterTopologyChangesIT.java [hunk 19]
```diff
@@ -91,20 +92,34 @@ public void setup() throws Exception
                 .startCluster();
     }
 
+    @After
+    public void cleanup()
+    {
+        cluster = null;
+        System.gc();
```
Comment (davidegrohmann) [inline enterprise/ha/src/test/java/org/neo4j/kernel/ha/ClusterTopologyChangesIT.java]: is this really needed?

File: enterprise/ha/src/test/java/org/neo4j/kernel/ha/ClusterTopologyChangesIT.java [hunk 20]
```diff
@@ -91,20 +92,34 @@ public void setup() throws Exception
...
+    }
+    
     @Test
     public void masterRejoinsAfterFailureAndReelection() throws Throwable
     {
+        System.out.println(repeat.getCount());
```
Comment (davidegrohmann) [inline enterprise/ha/src/test/java/org/neo4j/kernel/ha/ClusterTopologyChangesIT.java]: to be removed?

File: enterprise/ha/src/test/java/org/neo4j/kernel/ha/ClusterTopologyChangesIT.java [hunk 21]
```diff
@@ -78,6 +76,9 @@
     @Rule
     public final CleanupRule cleanup = new CleanupRule();
 
+    @Rule
+    public final RepeatRule repeat = new RepeatRule();
+
```
Comment (davidegrohmann) [inline enterprise/ha/src/test/java/org/neo4j/kernel/ha/ClusterTopologyChangesIT.java]: to be removed?

File: enterprise/ha/src/test/java/org/neo4j/ha/UniquenessConstraintValidationHAIT.java [hunk 22]
```diff
@@ -51,6 +58,9 @@
             testDirForTest( UniquenessConstraintValidationHAIT.class );
     public final @Rule OtherThreadRule<Void> otherThread = new OtherThreadRule<>();
 
+    public final @Rule
+    RepeatRule repeat = new RepeatRule();
+
```
Comment (davidegrohmann) [inline enterprise/ha/src/test/java/org/neo4j/ha/UniquenessConstraintValidationHAIT.java]: to be removed?

File: neo4j-home/data/dbms/authorization [hunk 23]
```diff
@@ -0,0 +1 @@
+neo4j:SHA-256,6192B103938C72EFB8FD8907764ADE6847F937574703356EC622FDE48E251F38,BC47DB8C9EF056ECC8D6C54F7033152A:password_change_required
```
Comment (davidegrohmann) [inline neo4j-home/data/dbms/authorization:1]: I guess this file shouldn't be part of the PR

File: data/dbms/auth [hunk 24]
```diff
@@ -0,0 +1 @@
+neo4j:SHA-256,427150500FAF017CCE86FF79C6E11700FD1EFCD8205CB905569FC67F70C52A87,C70FE30ADD59EE6D2047D654789D2B7F:password_change_required
```
Comment (davidegrohmann) [inline data/dbms/auth]: I guess this file shouldn't be part of the PR

**Comment by davidegrohmann** [comment]
Several small issues, but it looks good overall.

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/HighAvailabilityModeSwitcher.java (hunk 16 above)
Comment (rickardoberg) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/HighAvailabilityModeSwitcher.java]: It shuts down, but there's no restart or recovery, so this server is essentially bricked.

Comment (davidegrohmann) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/HighAvailabilityModeSwitcher.java]: perhaps add the explanation to the comment for future reference?

File: community/kernel/src/test/java/org/neo4j/kernel/lifecycle/LifeRule.java [hunk 25]
```diff
@@ -52,12 +53,13 @@ public void evaluate() throws Throwable
                         start();
                     }
                     base.evaluate();
+                    life.shutdown();
```
Comment (MishaDemianenko) [inline community/kernel/src/test/java/org/neo4j/kernel/lifecycle/LifeRule.java:56]: We used `start()` method before `base.evaluate()`. Can we change `life.shutdown()` to `shutdown()` here and in catch block?

File: enterprise/cluster/src/main/java/org/neo4j/cluster/com/NetworkReceiver.java [hunk 26]
```diff
@@ -234,14 +246,21 @@ private URI getURI( InetSocketAddress address ) throws URISyntaxException
...
-            uri = CLUSTER_SCHEME + "://" + address.getAddress().getHostAddress()+":"+address.getPort(); // Socket.toString() already prepends a /
+        {
+            uri = CLUSTER_SCHEME + "://" + address.getAddress().getHostAddress() + ":" + address.getPort(); // Socket
+        }
+            // .toString() already prepends a /
 
         // Add name if given
-        if (config.name() != null)
-            uri += "/?name="+config.name();
+        if ( config.name() != null )
```
Comment (MishaDemianenko) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/com/NetworkReceiver.java:260]: Should we also check for empty string here?

File: enterprise/cluster/src/main/java/org/neo4j/cluster/client/Clusters.java [hunk 27]
```diff
@@ -203,14 +201,7 @@ public boolean isFullHaMember()
...
         {
-            try
-            {
-                return InetAddress.getLocalHost().getHostAddress();
-            }
-            catch ( UnknownHostException e )
-            {
-                throw new RuntimeException( e );
-            }
+            return "127.0.0.1";
```
Comment (MishaDemianenko) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/client/Clusters.java:204]: Does it really safe to assume our localhost is always 127.0.0.1?

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/ProposerMessage.java [hunk 28]
```diff
@@ -61,9 +64,22 @@ public Object getValue()
...
+            Object toStringValue = value;
+            if (toStringValue instanceof Payload )
+            {
+                try
+                {
+                    toStringValue = new AtomicBroadcastSerializer( new ObjectStreamFactory(), new ObjectStreamFactory() ).receive( (Payload) toStringValue);
+                }
+                catch ( Throwable e )
+                {
+                    e.printStackTrace();
```
Comment (MishaDemianenko) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/ProposerMessage.java:76]: Can we make way how we handle exceptions during `toString()` method call consistent? Now its either empty string either stack trace or nothing at all.

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/InMemoryAcceptorInstanceStore.java [hunk 29]
```diff
@@ -29,18 +31,21 @@
         implements AcceptorInstanceStore
 {
     private final Map<InstanceId, AcceptorInstance> instances;
+    private final BlockingQueue<InstanceId> currentInstances;
```
Comment (MishaDemianenko) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/InMemoryAcceptorInstanceStore.java:34]: Can you please add description what is general idea of this class and why we need instances, currentInstances fields and why they can be out of sync - as example after clear method call

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/ProposerState.java [hunk 30]
```diff
@@ -83,7 +83,8 @@ public ProposerState handle( ProposerContext context,
                                     + " rejected from " + message.getHeader( Message.FROM ) + " with ballot "
                                     + rejectPropose.getBallot() );
 
-                            if ( instance.isState( PaxosInstance.State.p1_pending ) )
+                            if ( instance.isState( PaxosInstance.State.p1_pending ) &&
+                                    instance.getBallot() < rejectPropose.getBallot()) // Ignore multiple rejects on same prepare
```
Comment (MishaDemianenko) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/ProposerState.java:87]: Can you please check this at your test in `ProposerStateTest`?
